local/man/man3/ema.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.31)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ema 3"
.TH ema 3 "2016-02-03" "perl v5.22.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ema \- perl library for Exponential Moving Average
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use ema;
\&
\& $e = new ema (0.5, 0);
\& printf "%.3f", $e\->filter (1);
\& printf "%.3f", $e\->filter (1);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBema\fR library implements an exponential moving average,
effectively the discrete version of a \s-1RC\s0 low-pass filter.  All input
and return values are floating point.
.PP
See <https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average>
.ie n .IP "$e = new ema ($alpha, $initial_value)" 4
.el .IP "\f(CW$e\fR = new ema ($alpha, \f(CW$initial_value\fR)" 4
.IX Item "$e = new ema ($alpha, $initial_value)"
Construct a new \fBema\fR object with the given \fBalpha\fR, and optional
\&\fBinitial_value\fR.  If no \fBinitial_value\fR is given, zero will be used.
.Sp
The object contains all the information necessary to compute a moving average
on one set of data.  If you want to filter multiple sets of data, create
multiple objects.
.Sp
The value of \fBalpha\fR controls how fast the filter converges on the new value.
\&\fBAlpha\fR should be small, between 0 and 1.  Smaller values of \fBalpha\fR will
take longer to converge and will effectively filter more.  In terms of classic
\&\s-1RC\s0 filters, the time constant, T = 1/\fBalpha\fR.  After 1/\fBalpha\fR samples, the
filter will have converged 1\-exp(\-1)=63.2% of the way to the new input value.
.Sp
The cutoff frequency of the filter is 1/T radians/second.  As a
period, this will be 2Pi*T samples.  At this frequency, the filter
passes half the power, or 1/\fIsqrt\fR\|(2) = 0.707 of the amplitude.
.ie n .IP "$y = $e\->value ($x)" 4
.el .IP "\f(CW$y\fR = \f(CW$e\fR\->value ($x)" 4
.IX Item "$y = $e->value ($x)"
Advance the moving average by including a new input value \fB\f(CB$x\fB\fR and return the
next filtered value.
.ie n .IP "$e\->y" 4
.el .IP "\f(CW$e\fR\->y" 4
.IX Item "$e->y"
Return the current value of the filter without advancing it.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIstats\fR\|(3), \fIleastsquares\fR\|(3)
.SH "AUTHOR"
.IX Header "AUTHOR"
Robert Dowling <\fIRPNCalcN@gmail.com\fR>.


local/lib/ema.pm
package ema;

# Exponential Moving Average
#
'(C) Copyright 2016 Robert Dowling.';
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
# Usage
#
# $e = new ema (0.5, 0);
# printf "%.3f", $e->filter (1);
# printf "%.3f", $e->filter (1);
# printf "%.3f", $e->filter (1);

require Exporter;
our @ISA = qw(Exporter);
# our @EXPORT = qw();

sub new
{
    my ($type, $alpha, $y) = @_;
    my $self = {};
    $self->{alpha} = $alpha;
    $self->{y} = $y;
    bless $self, $type;
    $self;
}

# Accept a new value and return a new filtered value
sub filter ($)
{
    my ($self, $x) = @_;
    $self->{y} += ($x - $self->{y}) * $self->{alpha};
}

# Return the last filtered value
sub y ()
{
    my $self = shift;
    $self->{y};
}

use constant PI => 3.1415926535;

# Return the cutoff frequency of this filter as a period
sub cutoff_period ()
{
    my $self = shift;
    2 * PI / $self->{alpha};
}

# Return the frequency of this filter relative to the sample rate
sub cutoff_frequency ($)
{
    my ($self, $T) = @_;
    $T / $self->cutoff_period;
}

# Return the phase delay in radians at the given frequency and sample rate
sub phase_delay ($$)
{
    my ($self, $f, $T) = @_;
    atan2($f,$self->cutoff_frequency($T));
}

# Return the phase delay in samples of the given frequency and sample rate
sub phase_delay_samples ($$)
{
    my ($self, $f, $T) = @_;
    $T/$f*$self->phase_delay($f,$T)/(2*PI);
}

1;
__END__

=head1 NAME

ema - perl library for Exponential Moving Average

=head1 SYNOPSIS

 use ema;

 $e = new ema (0.5, 0);
 printf "%.3f", $e->filter (1);
 printf "%.3f", $e->filter (1);

=head1 DESCRIPTION

The B<ema> library implements an exponential moving average,
effectively the discrete version of a RC low-pass filter.  All input
and return values are floating point.

See L<https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average>

=over

=item $e = new ema ($alpha, $initial_value)

Construct a new B<ema> object with the given B<alpha>, and optional
B<initial_value>.  If no B<initial_value> is given, zero will be used.

The object contains all the information necessary to compute a moving average
on one set of data.  If you want to filter multiple sets of data, create
multiple objects.

The value of B<alpha> controls how fast the filter converges on the new value.
B<Alpha> should be small, between 0 and 1.  Smaller values of B<alpha> will
take longer to converge and will effectively filter more.  In terms of classic
RC filters, the time constant, T = 1/B<alpha>.  After 1/B<alpha> samples, the
filter will have converged 1-exp(-1)=63.2% of the way to the new input value.

The cutoff frequency of the filter is 1/T radians/second.  As a
period, this will be 2Pi*T samples.  At this frequency, the filter
passes half the power, or 1/sqrt(2) = 0.707 of the amplitude.

The phase delay going through the filter is -atan(omega/omega_0)

=item $y = $e->value ($x)

Advance the moving average by including a new input value B<$x> and return the
next filtered value.

=item $e->y

Return the current value of the filter without advancing it.

=item $samples = $e->cutoff_period;

=item $frequency = $e->cutoff_frequency($sample_rate);

Return the filter cutoff in B<$samples> or as the B<$frequency> for a
given B<$sample_rate>.

=item $radians = $e->phase_delay ($freq, $sample_rate);

=item $samples = $e->phase_delay_samples ($freq, $sample_rate);

Return the phase delay in B<radians> or in B<$samples> for the given
B<$frequency> and B<$sample_rate>.

=back

=head1 SEE ALSO

stats(3), leastsquares(3)

=head1 AUTHOR

Robert Dowling <F<RPNCalcN@gmail.com>>.

=cut
a
